src/utils/survey.ts

```diff
@@ -0,0 +1,238 @@
+
+const localize = nls.loadMessageBundle();
+let isCandidate: boolean | undefined = undefined;
+// this will be set to true if the user interacts with the prompt,
+// from there on whole logic will be disarmed for the rest of the session including scoring
+let wasPromptedInSession: boolean = false;
+
+const usageScoreByExperience: Record<ExperienceKind, number> = {
+    [ExperienceKind.Mongo]: 0,
+    [ExperienceKind.NoSQL]: 0,
+};
+
+export function countExperienceUsageForSurvey(experience: ExperienceKind, score: UsageImpact | number): void {
+    if (DISABLE_SURVEY || wasPromptedInSession) {
+        return;
+    }
+    usageScoreByExperience[experience] += score;
+}
+
+export async function promptAfterActionEventually(
+    experience: ExperienceKind,
+    score: UsageImpact | number,
+    triggerAction?: string,
+): Promise<void> {
+    if (DISABLE_SURVEY || wasPromptedInSession) {
+        return;
+    }
+
+    countExperienceUsageForSurvey(experience, score);
+
+    const { fullScore, highestExperience } = (
+        Object.entries(usageScoreByExperience) as [ExperienceKind, number][]
+    ).reduce(
+        (acc, entry) => {
+            acc.fullScore += entry[1];
+            if (entry[1] > acc.highestExperience[1]) {
+                acc.highestExperience = entry;
+            }
+            return acc;
+        },
+        { fullScore: 0, highestExperience: [ExperienceKind.Mongo, 0] as [ExperienceKind, number] }, // initial value
+    );
+
+    if (fullScore >= 100) {
+        await surveyPromptIfCandidate(highestExperience[0], triggerAction);
+    }
+}
+
+export async function getIsSurveyCandidate(): Promise<boolean> {
+    if (DISABLE_SURVEY) {
+        return false;
+    }
+    if (isCandidate === undefined) {
+        await initSurvey();
+    }
+    return isCandidate ?? false;
+}
+
+async function initSurvey(): Promise<void> {
+    await callWithTelemetryAndErrorHandling('survey.init', async (context: IActionContext) => {
+        if (DEBUG_ALWAYS_PROMPT) {
+            context.telemetry.properties.isCandidate = (isCandidate = true).toString();
+            await ext.context.globalState.update(IS_CANDIDATE_KEY, isCandidate);
+            return;
+        }
+
+        // Prompt only for English locales
+        if (PROMPT_ENGLISH_ONLY && env.language !== 'en' && !env.language.startsWith('en-')) {
+            context.telemetry.properties.isCandidate = (isCandidate = false).toString();
+            return;
+        }
+
+        // Prompt only once per major/minor version,
+        // SKIP_VERSION_KEY will be set to the version that was prompted already and the user either clicked "Don't Ask Again" or opened the survey
+        const extensionVersion = (ext.context.extension.packageJSON as { version: string }).version;
+        const extensionSemVer = semver.parse(extensionVersion);
+        const skipVersion = semver.parse(ext.context.globalState.get(SKIP_VERSION_KEY, ''));
```

Comment:
for state there is a special service `SettingsService`. I would like to avoid any kind of usage `ext`. Since no one controls this global variable.

Reply:
https://github.com/microsoft/vscode-cosmosdb/issues/2597